packages/flutter_tools/lib/src/windows/migrations/version_migration.dart

```diff
@@ -0,0 +1,114 @@
+'\n'
+'# Disable Windows macros that collide with C++ standard library functions.\n'
+'target_compile_definitions(\${BINARY_NAME} PRIVATE "NOMINMAX")\n';
+
+const String _resourceFileBefore =
+'#ifdef FLUTTER_BUILD_NUMBER\n'
+'#define VERSION_AS_NUMBER FLUTTER_BUILD_NUMBER\n'
+'#else\n'
+'#define VERSION_AS_NUMBER 1,0,0\n'
+'#endif\n'
+'\n'
+'#ifdef FLUTTER_BUILD_NAME\n'
+'#define VERSION_AS_STRING #FLUTTER_BUILD_NAME\n'
+'#else\n'
+'#define VERSION_AS_STRING "1.0.0"\n'
+'#endif\n';
+const String _resourceFileAfter =
+'#if defined(FLUTTER_VERSION_MAJOR) && defined(FLUTTER_VERSION_MINOR) && defined(FLUTTER_VERSION_PATCH) && defined(FLUTTER_VERSION_BUILD)\n'
+'#define VERSION_AS_NUMBER FLUTTER_VERSION_MAJOR,FLUTTER_VERSION_MINOR,FLUTTER_VERSION_PATCH,FLUTTER_VERSION_BUILD\n'
+'#else\n'
+'#define VERSION_AS_NUMBER 1,0,0,0\n'
+'#endif\n'
+'\n'
+'#if defined(FLUTTER_VERSION)\n'
+'#define VERSION_AS_STRING FLUTTER_VERSION\n'
+'#else\n'
+'#define VERSION_AS_STRING "1.0.0"\n'
+'#endif\n';
+
+// Flutter should set the Windows app's version information.
+// See https://github.com/flutter/flutter/issues/73652.
+class VersionMigration extends ProjectMigrator {
+  VersionMigration(WindowsProject project, super.logger)
+    : _cmakeFile = project.runnerCmakeFile, _resourceFile = project.runnerResourceFile;
+
+  final File _cmakeFile;
+  final File _resourceFile;
+
+  @override
+  void migrate() {
+    if (!_cmakeFile.existsSync()) {
+      logger.printTrace('windows/runner/CMakeLists.txt file not found, skipping version migration');
+      return;
+    }
+
+    if (!_resourceFile.existsSync()) {
+      logger.printTrace('windows/runner/Runner.rc file not found, skipping version migration');
+      return;
+    }
+
+    // Migrate the windows/runner/CMakeLists.txt file.
+    final String originalCmakeContents = _cmakeFile.readAsStringSync();
+    final String newCmakeContents = _replaceFirstWindows(
+      originalCmakeContents,
+      _cmakeFileBefore,
+      _cmakeFileAfter,
+    );
+    if (originalCmakeContents != newCmakeContents) {
+      logger.printStatus('windows/runner/CMakeLists.txt does not define version information, updating.');
+      _cmakeFile.writeAsStringSync(newCmakeContents);
+    }
+
+    // Migrate the windows/runner/Runner.rc file.
+    final String originalResourceFileContents = _resourceFile.readAsStringSync();
+    final String newResourceFileContents = _replaceFirstWindows(
+      originalResourceFileContents,
+      _resourceFileBefore,
+      _resourceFileAfter,
+    );
+    if (originalResourceFileContents != newResourceFileContents) {
+      logger.printStatus(
+        'windows/runner/Runner.rc does not define use Flutter version information, updating.',
+      );
+      _resourceFile.writeAsStringSync(newResourceFileContents);
+    }
+  }
+}
+
+String _replaceFirstWindows(String originalContents, String before, String after) {
```

Comment:
Even though it's private, consider adding a doc comment stating what this does. The name sounds like it might have to do with replacing window resources of some sort (if such a thing exists).

Do we already have a `_replaceFirst`? I'd be tempted to just name it that, since that's what it does. That it handles windows line-endings is more of an implementation detail.

Reply:
Updated! Let me know if you have additional feedback here!